SignalServiceKit/src/Messages/Interactions/TSOutgoingMessage.m

```diff
@@ -60,22 +62,112 @@ - (instancetype)initWithCoder:(NSCoder *)coder
         if (!_attachmentFilenameMap) {
             _attachmentFilenameMap = [NSMutableDictionary new];
         }
-        
-        // Migrate message state.
-        if (_messageState == TSOutgoingMessageStateSent_OBSOLETE) {
-            _messageState = TSOutgoingMessageStateSentToService;
-        } else if (_messageState == TSOutgoingMessageStateDelivered_OBSOLETE) {
-            _messageState = TSOutgoingMessageStateSentToService;
-            _wasDelivered = YES;
-        }
-        if (!_sentRecipients) {
-            _sentRecipients = [NSArray new];
+
+        if (!self.recipientStateMap) {
+            [self migrateRecipientStateMapWithCoder:coder];
+            OWSAssert(self.recipientStateMap);
         }
     }
 
     return self;
 }
 
+- (void)migrateRecipientStateMapWithCoder:(NSCoder *)coder
```

Comment:
The heart of this PR is migrating from all of the conflicting, redundant or ambiguous state we have to a clear, streamlined model of recipient state.

Reply:
👏 👏 👏